src/chainstate/burn/distribution.rs

```diff
@@ -0,0 +1,827 @@
+            let block_header_hash_160 = Hash160::from_sha256(&burn_sample[i].candidate.block_header_hash.as_bytes());
+            let vrf_hex = ECVRF_public_key_to_hex(&burn_sample[i].key.public_key);
+            assert!(burn_index.get(&(vrf_hex.clone(), block_header_hash_160)).is_none());
+
+            burn_index.insert((vrf_hex, block_header_hash_160), i);
+        }
+
+        // match user burns to the leaders
+        for user_burn in user_burns {
+            let block_idx_opt = burn_index.get(&(ECVRF_public_key_to_hex(&user_burn.public_key), user_burn.block_header_hash_160));
+            match block_idx_opt {
+                Some(ref_block_idx) => {
+                    // user burn matches a (key, block) pair committed to in this block
+                    let idx = *ref_block_idx;
+                    burn_sample[idx].burns += (user_burn.burn_fee as u128);
+                    burn_sample[idx].user_burns.push(user_burn);
+                },
+                None => {
+                    info!("User burn {} ({},{}) of {} for key={}, block={} has no matching block commit",
+                          &user_burn.txid.to_hex(), user_burn.block_number, user_burn.vtxindex, user_burn.burn_fee,
+                          ECVRF_public_key_to_hex(&user_burn.public_key), &user_burn.block_header_hash_160.to_hex());
+                    continue;
+                }
+            };
+        }
+    }
+
+    /// Calculate the ranges between 0 and 2**256 - 1 over which each point in the burn sample
+    /// applies, so we can later select which block to use.
+    fn make_sortition_ranges(burn_sample: &mut Vec<BurnSamplePoint<A, K>>) -> () {
+        if burn_sample.len() == 0 {
+            // empty sample
+            return;
+        }
+        if burn_sample.len() == 1 {
+            // sample that covers the whole range 
+            burn_sample[0].range_start = Uint256::zero();
+            burn_sample[0].range_end = Uint256::max();
+            return;
+        }
+
+        // total burns for valid blocks?
+        // NOTE: this can't overflow -- there's no way we get that many (u64) burns
+        let total_burns_u128: u128 = burn_sample
+            .iter()
+            .fold(0u128, |mut total_burns, sample| {
+                total_burns = total_burns + (sample.burns as u128);
+                total_burns
+            });
+
+        let total_burns = Uint512::from_u128(total_burns_u128);
+
+        // determine range start/end for each sample.
+        // Use fixed-point math on an unsigned 512-bit number -- 
+        //   * the upper 256 bits are the integer
+        //   * the lower 256 bits are the fraction
+        // These range fields correspond to ranges in the 32-byte hash space
+        let mut burn_acc = Uint512::from_u128(burn_sample[0].burns);
+
+        burn_sample[0].range_start = Uint256::zero();
+        burn_sample[0].range_end = ((Uint512::from_uint256(&Uint256::max()) * burn_acc) / total_burns).to_uint256();
```

Comment:
in this line (and the one below), shouldn't this be `1+Uint512::from_uint256(&Uint256::max())`? The integer "1" should be unrepresantable as a u256.

Reply:
I don't think so, but maybe you can check my reasoning?

Conceptually, the `range_start` and `range_end` represent numbers in the open interval [0, 1).  We're partitioning this interval into contiguous slices allocated to each block candidate, such that a candidate's slice size is proportional to the fraction burned for the candidate.  The first candidate's interval is `[0, burn_sample[0].burns / burn_total)`.  The second candidate's interval is `[burn_sample[0].burns / burn_total, (burn_sample[0].burns + burn_sample[1].burns) / burn_total)`, etc.  Because we're not using real numbers, the value of `sup [0, 1)` (i.e. the largest value in the last candidate's interval) would simply be the maximum possible value you can stuff into a 256-bit integer.  This is why I'm using `Uint256::max()` instead of `1 << 256`.

Reply:
ah, I see now. since burn_accum / burn_total can never be > 1,  all that you're really trying to do here is take (2^256-1) and multiply by (burn_accum/burn_total), but you use u512s for increased accuracy (as opposed to say, dividing by the burn_total first, and then multiplying by burn_accum).